main.nf

```diff
@@ -140,392 +165,666 @@ process fastqc {
-    bowtie \
-        -p 2 \
-        -t \
-        -n 0 \
-        -l 15 \
-        -e 99999 \
-        -k 200 \
-        --best \
-        -S \
-        --chunkmbs 2048 \
-        $bowtie_miRBase_mature \
-        $reads \
-        | samtools view -bS > \${f}.aligned_miRBase_mature.bam
-    """
-}
+    /* Note! Only read 1 is used for alignment. */
+
+    single = reads instanceof Path
+
+    if(single) {
+      """
+      f='$reads';f=(\$f);input=\${f[0]};f=\${input%.gz};f=\${f%.fastq};f=\${f%.fq};f=\${f%.R1_val_1*};f=\${f%.R2_val_2*};f=\${f%_val_1};f=\${f%_val_2};f=\${f%_trimmed};f=\${f%_1}
+      prefix=\$f
+
+      bowtie $mature -q <(zcat \$input) -v 0 -k 10 -S \${prefix}.mature.sam --un \${prefix}.mature_unmapped.fq
 
+      gzip \${prefix}.mature_unmapped.fq
+      """
+    } else {
+      """
+      f='$reads';f=(\$f);input=\${f[0]};f=\${input%.gz};f=\${f%.fastq};f=\${f%.fq};f=\${f%.R1_val_1*};f=\${f%.R2_val_2*};f=\${f%_val_1};f=\${f%_val_2};f=\${f%_trimmed};f=\${f%_1}
+      prefix=\$f
+
+      bowtie $mature -q <(zcat \$input) -v 0 -k 10 -S \${prefix}.mature.sam --un \${prefix}.mature_unmapped.fq
+
+      gzip \${prefix}.mature_unmapped.fq
+      """
+    }
+}
 
 /*
- * STEP 4 - Bowtie against miRBase hairpin RNA
+ * STEP 3.2 - Bowtie against miRBase mature miRNA post-alignment processing
  */
-process bowtie_miRBase_hairpin {
-    
+
+process bowtie_miRBase_mature_postalignment {
+
+    tag "$miRBase_mature_sam"
+
     module 'bioinfo-tools'
-    module 'bowtie'
     module 'samtools'
-    
-    memory '2 GB'
-    time '6h'
+    module 'BEDTools'
+
+    cpus 2
+    memory { 16.GB * task.attempt }
+    time { 120.h * task.attempt }
+    errorStrategy { task.exitStatus == 143 ? 'retry' : 'terminate' }
+    maxRetries 3
+    maxErrors '-1'
+
+    publishDir "${params.outdir}/bowtie_miRBase_mature", mode: 'copy'
 
     input:
-    file(reads:'*') from trimmed_reads
-    
+    file miRBase_mature_sam
+
     output:
-    file '*.bam' into miRBase_hairpin_bam
-    
+    file '*.sorted.bam' into miRBase_mature_bam
+    file '*.sorted.bam.bai' into miRBase_mature_bai
+    file '*.count' into miRBase_mature_count
+
     script:
     """
-    f='$reads';f=(\$f);f=\${f[0]};f=\${f%.gz};f=\${f%.fastq};f=\${f%.fq};f=\${f%_val_1};f=\${f%_trimmed};f=\${f%_1}
-    bowtie \
-        -p 2 \
-        -t \
-        -n 0 \
-        -l 15 \
-        -e 99999 \
-        -k 200 \
-        --best \
-        -S \
-        --chunkmbs 2048 \
-        $bowtie_miRBase_hairpin \
-        $reads \
-        | samtools view -bS > \${f}.aligned_miRBase_hairpin.bam
+    f='$miRBase_mature_sam';f=(\$f);f=\${f[0]};f=\${f%.sam}
+    prefix=\$f
+
+    samtools view -bS \${prefix}.sam > \${prefix}.bam
+    samtools sort \${prefix}.bam \${prefix}.sorted
+    samtools index \${prefix}.sorted.bam
+    samtools idxstats \${prefix}.sorted.bam > \${prefix}.count
+
+    rm \${prefix}.sam
```

Comment:
No need for this - wrapper script will delete these anyway if they're not copied to the results directory. Better to keep intermediate files I think.